internal/resource/generic_fetcher.go

```diff
@@ -0,0 +1,304 @@
+					op.FetchOptions.ExpectedSum, Err = hex.DecodeString(expectedSumString)
+					if Err != nil {
+						l.Err("%s: error parsing verification string - %q: %v", msg, expectedSumString, Err)
+						return
+					}
+				}
+
+			case ParsingHTTPHeaders:
+				if op.ResCfg != nil && op.ResCfg.HTTPHeaders != nil && len(op.ResCfg.HTTPHeaders) > 0 {
+					op.FetchOptions.Headers, Err = op.ResCfg.HTTPHeaders.Parse()
+					if Err != nil {
+						l.Err("%s: error parsing http headers - %v: %v", msg, op.ResCfg.HTTPHeaders, Err)
+						return
+					}
+				}
+
+			case CreatingTempFile:
+				if op.File == nil {
+					continue
+				}
+				Err = os.MkdirAll(filepath.Dir(op.File.Node.Path), DefaultDirectoryPermissions)
+				if Err != nil {
+					l.Err("%s: error making dir for file - %v", msg, Err)
+					return
+				}
+				// Create a temporary file in the same directory to ensure it's on the same filesystem
+				tmp, Err = ioutil.TempFile(filepath.Dir(op.File.Node.Path), "tmp")
+				if Err != nil {
+					l.Err("%s: error creating tmp file - %v", msg, Err)
+					return
+				}
+				defer tmp.Close()
+
+			case ChangingPermissions:
+				if op.File == nil {
+					continue
+				}
+				// ioutil.TempFile defaults to 0600
+				Err = tmp.Chmod(DefaultFilePermissions)
+				if Err != nil {
+					l.Err("%s: error changing file permissions - %q: %v", msg, tmp.Name(), Err)
+					return
+				}
+				// sometimes the following line will fail (the file might be renamed),
+				// but that's ok (we wanted to keep the file in that case).
+				defer os.Remove(tmp.Name())
+
+			case Fetching:
+				if op.ResCfg != nil && op.ResCfg.Compression != nil {
+					op.FetchOptions.Compression = *op.ResCfg.Compression
+				}
+				if op.File == nil {
+					Ret.Cfg, Err = f.FetchToBuffer(*Ret.Uri, op.FetchOptions, abort)
+				} else {
+					Err = f.Fetch(*Ret.Uri, tmp, op.FetchOptions, abort)
+				}
+				if Err != nil {
+					l.Err("%s: failed - %s", msg, Err)
+					return
+				}
+
+			case Appending:
+				if op.File == nil {
+					continue
+				}
+				if op.Append {
+					var finfo os.FileInfo
+					// Make sure that we're appending to a file
+					finfo, Err = os.Lstat(op.File.Node.Path)
+					switch {
+					case os.IsNotExist(Err):
+						// No problem, we'll create it.
+						break
+					case Err != nil:
+						return
+					default:
+						if !finfo.Mode().IsRegular() {
+							Err = fmt.Errorf("can only append to files: %q", op.File.Node.Path)
+							return
+						}
+					}
+
+					// Open with the default permissions, we'll chown/chmod it later
+					var targetFile *os.File
+					targetFile, Err = os.OpenFile(op.File.Node.Path, os.O_WRONLY|os.O_APPEND|os.O_CREATE, DefaultFilePermissions)
+					if Err != nil {
+						return
+					}
+					defer targetFile.Close()
+
+					if _, Err = tmp.Seek(0, io.SeekStart); Err != nil {
+						return
+					}
+					if _, Err = io.Copy(targetFile, tmp); Err != nil {
+						return
+					}
```

Comment:
Is there something that would prevent an edge case where multiple fetches take a similar / the same amount of time to enter the `Appending` stage and multiple copies of the content are appended into a file?

Reply:
added a semaphore